src/ccs/cccs.rs

```diff
@@ -0,0 +1,279 @@
+};
+use bitvec::vec;
+use core::{cmp::max, marker::PhantomData};
+use ff::{Field, PrimeField};
+use flate2::{write::ZlibEncoder, Compression};
+use itertools::concat;
+use rayon::prelude::*;
+use serde::{Deserialize, Serialize};
+use sha3::{Digest, Sha3_256};
+use std::ops::{Add, Mul};
+use std::sync::Arc;
+
+use super::util::compute_sum_Mz;
+use super::util::virtual_poly::VirtualPolynomial;
+use super::{CCSShape, CCSWitness};
+
+/// A type that holds the shape of a Committed CCS (CCCS) instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct CCCSShape<G: Group> {
+  // Sequence of sparse MLE polynomials in s+s' variables M_MLE1, ..., M_MLEt
+  pub(crate) M_MLE: Vec<MultilinearPolynomial<G::Scalar>>,
+
+  pub(crate) ccs: CCSShape<G>,
+}
+
+/// A type that holds a CCCS instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct CCCSInstance<G: Group> {
+  // Commitment to a multilinear polynomial in s' - 1 variables
+  pub(crate) C: Commitment<G>,
+
+  // $x in F^l$
+  pub(crate) x: Vec<G::Scalar>,
+}
+
+impl<G: Group> CCCSShape<G> {
+  // Computes q(x) = \sum^q c_i * \prod_{j \in S_i} ( \sum_{y \in {0,1}^s'} M_j(x, y) * z(y) )
+  // polynomial over x
+  pub fn compute_q(&self, z: &Vec<G::Scalar>) -> Result<VirtualPolynomial<G::Scalar>, NovaError> {
+    let z_mle = dense_vec_to_mle::<G::Scalar>(self.ccs.s_prime, z);
+    if z_mle.get_num_vars() != self.ccs.s_prime {
+      // this check if redundant if dense_vec_to_mle is correct
+      return Err(NovaError::VpArith);
+    }
+
+    // Using `fold` requires to not have results inside. So we unwrap for now but
+    // a better approach is needed (we ca just keep the for loop otherwise.)
+    Ok(
+      (0..self.ccs.q).fold(VirtualPolynomial::<G::Scalar>::new(self.ccs.s), |q, idx| {
+        let mut prod = VirtualPolynomial::<G::Scalar>::new(self.ccs.s);
+
+        for &j in &self.ccs.S[idx] {
+          let sum_Mz = compute_sum_Mz::<G>(&self.M_MLE[j], &z_mle);
+
+          // Fold this sum into the running product
+          if prod.products.is_empty() {
+            // If this is the first time we are adding something to this virtual polynomial, we need to
+            // explicitly add the products using add_mle_list()
+            // XXX is this true? improve API
+            prod
+              .add_mle_list([Arc::new(sum_Mz)], G::Scalar::ONE)
+              .unwrap();
+          } else {
+            prod.mul_by_mle(Arc::new(sum_Mz), G::Scalar::ONE).unwrap();
+          }
+        }
+        // Multiply by the product by the coefficient c_i
+        prod.scalar_mul(&self.ccs.c[idx]);
+        // Add it to the running sum
+        q.add(&prod)
+      }),
+    )
+  }
+
+  /// Computes Q(x) = \hat eq(beta, x) * q(x)
+  ///               = \hat eq(beta, x) * \sum^q c_i * \prod_{j \in S_i} ( \sum_{y \in {0,1}^s'} M_j(x, y) * z(y) )
+  /// polynomial over x
+  pub fn compute_Q(
```

Comment:
Is this necessary? I ask because when running the sum-check protocol on a mutivariate polynomial, it is typically implemented by keeping the sum-check polynomial (say g) in "decomposed" form i.e., using the multilinear polynomials that when multiplied together in some way, we get g, but neither the prover nor the verifier materializes the polynomial g. The sum-check messages are directly computed using individual multilinear polynomials. 

I'll keep reading the rest of the code to see how this method gets used.

Note that the sum-check works by keeping the sum-check polynomial (e.g.,